grpclb/src/main/java/io/grpc/grpclb/CachedSubchannelPool.java

```diff
@@ -34,32 +35,46 @@
  * A {@link SubchannelPool} that keeps returned {@link Subchannel}s for a given time before it's
  * shut down by the pool.
  */
+// TODO(creamsoup) address to subchannel is not 1:1 mapping because subchannel can update its
+//  address. Use Multimap and ForwardingSubchannel (override updateAddresses) to maintain most up to
```

Comment:
There should be no need for a multimap. When a subchannel is added to the pool it should no longer be used, which means updateAddresses shouldn't be called. When a subchannel is added this code can just call `subchannel.getAddresses()`/`getAllAddresses()` to add the subchannel to the `cache` map.

Reply:
i think it is a bit weird that the pool has no idea about the subchannels it created. i agree that the updateAddresses shouldn't be called externally, but nothing prevent it from doing that. And it seems like blocking it by throwing exception is not an option due to the hierarchical lb policy.
My point is the pool should know the status of subchannels it owns (ownership of the subchannel should be on the pool not the caller). currently, user of the pool can return a subchannel created elsewhere, and the pool won't complain. also, `clear` just shutdowns already returned subchannels and doesn't care about subchannels currently used.
if it was a SubchannelFactory it might look okay.

Reply:
It is very normal for pools dealing with exclusive objects have no tracking of the objects they hand out. "Exclusive" meaning that there can only be one user of the object at a time, like DB connections and the like. If the pool does keep track of the objects handed out, it is generally for resource limiting reasons (like a DB pool maxing out at 5 connections).

> And it seems like blocking it by throwing exception is not an option due to the hierarchical lb policy.

I don't understand this. The pool could wrap the returned subchannels and override getAddresses(). When the subchannel is returned to the pool the pool sets a `throwOnAddresses = true` variable in the subchannel that `getAddresses()` checks. When the subchannel is taken from the pool, `throwOnAddresses = false` is set before the pool returns the subchannel.

> My point is the pool should know the status of subchannels it owns (ownership of the subchannel should be on the pool not the caller)

I don't think so. And it clearly makes things harder. What benefit is there to having the pool track the subchannels it has handed out? If there is no benefit, why wouldn't we do it "the easy way?"

> currently, user of the pool can return a subchannel created elsewhere, and the pool won't complain

Fair. Although that is technically true of a couple different things related to subchannels (like which ones are returned from picks). This would be trivial to check in the future because we are guaranteed to wrap subchannel to handle `shutdown()`, so we could do an `instanceof` or (if we are really paranoid) a nested class instance check (`CachedSubchannelPool.this` within the subchannel).

> also, clear just shutdowns already returned subchannels and doesn't care about subchannels currently used.

That sounds perfect. Policies should shutdown/return the subchannels they created.

Reply:
okay thanks for the explanation. i think i went side ways by trying to get rid of the handleSubchannelState.